[REEF-1415] Set AsError to TaskExceptionMissingException on Fa…

afchung opened the pull request:
…iledTask for FailedEvaluators

This addressed the issue by
  * Documenting the FailedTask class.
  * Adding a helper function in ByteUtilities to check byte array IsNullOrEmpty.

JIRA:
  [REEF-1415](https://issues.apache.org/jira/browse/REEF-1415)

markusweimer commented on lang/cs/Org.Apache.REEF.Driver/Bridge/Events/FailedTask.cs:
```diff
@@ -87,16 +91,26 @@ public Optional<IActiveContext> GetActiveContext()
 
+        /// <summary>
+        /// Returns the Exception causing the Failed Task.
+        /// If the Exception was caused by a control flow error (start, stop, suspend), 
```
Please either use `<remarks>` or `<returns>` for these descriptions, as they are too long for a ToolTip :)

markusweimer commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskExceptionMissingException.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace Org.Apache.REEF.Driver.Task
+{
+    [Serializable]
```
Add a class level XML doc

markusweimer commented:
Done with a pass. Looks good overall. One design question came to mind: What if the FailedTask isn't a C# Task, but a Java task. REEF allows a C# Driver to orchestrate Java Evaluators, after all.

afchung commented:
@markusweimer is that feature actually feasible today through our provided APIs? If so, it's coverage we need to add to our suite of C# tests. If not I think we should open a JIRA and worry about it later.
I've also addressed your comments, please have another look. Thanks!

markusweimer commented:
We have an example which uses a Java Driver to orchestrate .NET Tasks. I don't think we have the reverse.

jwang98052 commented on lang/cs/Org.Apache.REEF.Driver/Bridge/Events/FailedTask.cs:
```diff
@@ -42,13 +41,19 @@ public FailedTask(IFailedTaskClr2Java failedTaskClr2Java)
-            Message = avroFailedTask.message ?? "No message in Failed Task.";
+
+            // Message can be overwritten in Java, if the C# Message is null and the Task failure is caused by an Evaluator failure.
+            Message = string.IsNullOrWhiteSpace(avroFailedTask.message) ? "No message in Failed Task." : avroFailedTask.message;
```
Would "No message in Failed Task from Evaluator" be better?

afchung replied:
I don't think it matters. All tasks run in Evaluators anyhow.

jwang98052 commented on lang/cs/Org.Apache.REEF.Driver/Bridge/Events/FailedTask.cs:
```diff
@@ -87,16 +89,29 @@ public Optional<IActiveContext> GetActiveContext()
 
-        private static Exception GetCause(byte[] serializedCause, string taskExceptionString)
+        private static Exception GetCause(
+            byte[] serializedCause, string originalTaskExceptionToString)
```
You can derive originalTaskExceptionToString from serializedCause, right? Then you don't need to pass the second parameter.

afchung replied:
`serializedCause` might not be able to be deserialized. In other words, the serialization implementation was correct but the deserialization implementation was incorrect.

jwang98052 commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskExceptionMissingException.cs:
```diff
@@ -0,0 +1,34 @@
+    /// before a Task is reported to be initialized.
+    /// </summary>
+    [Serializable]
+    public sealed class TaskExceptionMissingException : Exception
```
Do you need to override  public TaskExceptionMissingException (SerializationInfo info, StreamingContext context) to make it Serializable?

afchung replied:
Good catch.

jwang98052 commented:
I completed a pass.

afchung commented:
@markusweimer I've created REEF-1422 and added TODOs in the relevant parts.
@jwang98052 I've addressed your comments.

This is ready for another round of reviews. Thanks!

jwang98052 commented:
LGTM.

markusweimer commented:
LGTM, will merge
